Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -13,22 +13,183 @@
+                return
+            }
+
+            let callbackOpaque = Unmanaged<CallbackOpaquePointer>.fromOpaque(pointer).takeUnretainedValue()
+            let messageID = callbackOpaque.messageID
+            let producer = callbackOpaque.producer
+
+            Task {
+                await producer.deallocateCallbackOpaquePointer(pointer)
+                await producer.markMessageAsAcknowledged(messageID)
+            }
+        }
+
         self.topicConfig = topicConfig
+        self.client = try KafkaClient(type: .producer, config: config, logger: logger)
+        self.kafkaHandle = client.kafkaHandle
+        self.topicHandles = [:]
+        self.unacknowledgedMessages = [:]
+        self.receivedMessages = []
+        self.callbackOpaquePointers = []
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Kafka producer must be closed **explicitly**.
+    public func close() {
+        // Wait 10 seconds for outstanding messages to be sent and callbacks to be called
+        rd_kafka_flush(kafkaHandle, 10_000)
+
+        for (_, topicHandle) in self.topicHandles {
+            rd_kafka_topic_destroy(topicHandle)
+        }
+
+        callbackOpaquePointers.forEach {
+            _ = Unmanaged<CallbackOpaquePointer>.fromOpaque($0).takeRetainedValue()
+        }
+
+        self.pollTask.cancel()
+    }
+
+    /// Send messages to the Kafka cluster asynchronously, aka "fire and forget".
+    /// This function is non-blocking.
+    /// - Parameter message: The ``KafkaProducerMessage`` that is sent to the KafkaCluster.
+    public func sendAsync(message: KafkaProducerMessage) throws {
+        let topicHandle = createTopicHandleIfNeeded(topic: message.topic)
+
+        let keyBytes: [UInt8]?
+        if let key = message.key {
+            keyBytes = key.withUnsafeBytes { Array($0) }
+        } else {
+            keyBytes = nil
+        }
+
+        let callbackOpaque = CallbackOpaquePointer(self)
+        let pointer = UnsafeMutableRawPointer(Unmanaged.passRetained(callbackOpaque).toOpaque())
+        callbackOpaquePointers.append(pointer)
+
+        let responseCode = message.value.withUnsafeBytes { valueBuffer in
+
+            return rd_kafka_produce(
+                topicHandle,
+                message.partition.rawValue,
+                RD_KAFKA_MSG_F_COPY,
+                UnsafeMutableRawPointer(mutating: valueBuffer.baseAddress),
+                valueBuffer.count,
+                keyBytes,
+                keyBytes?.count ?? 0,
+                pointer
+            )
+        }
+
+        guard responseCode == 0 else {
+            throw KafkaError(rawValue: responseCode)
+        }
+
+        // TODO: can messages be acked before the message gets set? Should not be possible -> we are an actor
```

Comment:
Just to double-check, the question should rather be

`an messages be acked before the message callback gets set?`

However, I think the `actor` keeps us safe here